### PR TITLE
Add new language extensions for 9.2

### PR DESCRIPTION
**Pull request by emilypi**
---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

See: #7312

### CODE REVIEW

**Comment by emilypi** [comment]
@phadej does anything else need to happen to add support for new extensions?

**Review by fgaz — APPROVED** [review]

File: Cabal/src/Language/Haskell/Extension.hs
Comment (phadej) [inline Cabal/src/Language/Haskell/Extension.hs:859]: This is not `Extension`. This is `Language`.

**Review by phadej — COMMENTED** [review]
Yes, looks fine. 

Last time `cabal.vim` didn't exist, IIRC https://github.com/haskell/cabal/commit/753e006266a0ba5e83d6ccf7be1da14f6be598b1

I'd experiment with what `cabal-install-3.4` does when you give it an

```
library
  default-language: GHC2021
```

does it just pass it through (maybe with a warning that this is unknown language?)

That is how extension list works, IIRC, but maybe cabal is stricter with languages. I don't actually know.

File: Cabal/src/Language/Haskell/Extension.hs
Comment (emilypi) [reply Cabal/src/Language/Haskell/Extension.hs:859]: Shoot my bad - I was thinking extension for some reasno.

File: editors/vim/syntax/cabal.vim
Comment (phadej) [inline editors/vim/syntax/cabal.vim:189]: Still not an extension :)

**Comment by emilypi** [comment]
If this all looks good now, i'll merge when i get the 👍

File: Cabal/src/Distribution/Simple/GHC/ImplInfo.hs
Comment (phadej) [inline Cabal/src/Distribution/Simple/GHC/ImplInfo.hs:65]: `9.1` if you want this to work with pre-releases.

**Comment by bgamari** [comment]
This passes my local GHC validation. I have kicked off a full CI run [here](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/5506).

**Comment by emilypi** [comment]
@hvr any thoughts if you pop your head out?

**Comment by emilypi** [comment]
Confirmed via @bgamari

**Comment by fgaz** [comment]
It's too late to say this, but I think this branch needed a clean-up before merging. For example there are two commits that do the same thing (moving GHC2021 from the extensions to the languages) and the whole pr could have been one single commit since it can't be really split into multiple meaningful changes.

**Comment by emilypi** [comment]
I squashed these on merge @fgaz - there is only a single commit in master.  https://github.com/haskell/cabal/commit/bd20f1b2f09f3dfe735afd66c4779491479670ce

**Comment by fgaz** [comment]
Duh! Sorry for the noise then 😅

**Pull request merged**
